File: lib/request_importer.py
#!/usr/bin/env python3
"""
Request Importer - Import requests from various formats (Postman, cURL, JSON, YAML)
"""

import json
import re
import shlex
import yaml
from pathlib import Path
from typing import Dict, List, Optional, Any
from urllib.parse import urlparse


class RequestImporter:
    """Import GraphQL requests from various formats"""

    @staticmethod
    def _extract_operation_name(query: Optional[str]) -> Optional[str]:
        """Extract operation name from query if present."""
        if not query or not isinstance(query, str):
            return None

        op_match = re.search(r'(?:query|mutation|subscription)\s+(\w+)', query)
        if op_match:
            return op_match.group(1)
        return None

    @staticmethod
    def _normalize_variables(variables: Any) -> Any:
        """Normalize variables to a dict when possible."""
        if variables is None:
            return None
        if isinstance(variables, dict):
            return variables
        if isinstance(variables, str):
            stripped = variables.strip()
            if not stripped:
                return None
            try:
                parsed = json.loads(stripped)
                return parsed
            except Exception:
                return variables
        return variables

    @staticmethod
    def _extract_postman_requests(collection: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Extract requests from a Postman collection dictionary."""
        requests = []

        def extract_requests(item: Dict, folder_path: str = ""):
            """Recursively extract requests from collection items."""
            if item.get('request'):
                req = item['request']
                url_obj = req.get('url', {})

                # Build URL
                if isinstance(url_obj, str):
                    url = url_obj
                else:
                    protocol = url_obj.get('protocol', 'https')
                    host = url_obj.get('host', [])
                    path = url_obj.get('path', [])

                    if isinstance(host, list):
                        host = '.'.join(host)
                    if isinstance(path, list):
                        path = '/'.join(path)

                    url = f"{protocol}://{host}/{path}".replace('//', '/').replace(':/', '://')

                # Extract headers
                headers = {}
                for header in req.get('header', []):
                    if not header.get('disabled', False):
                        headers[header.get('key', '')] = header.get('value', '')

                # Extract body
                body = req.get('body', {})
                variables = None
                query = None

                if body.get('mode') == 'raw':
                    body_text = body.get('raw', '')
                    try:
                        body_json = json.loads(body_text)
                        query = body_json.get('query', '')
                        variables = RequestImporter._normalize_variables(body_json.get('variables'))
                    except Exception:
                        query = body_text
                elif body.get('mode') == 'graphql':
                    graphql = body.get('graphql', {})
                    query = graphql.get('query', '')
                    variables = RequestImporter._normalize_variables(graphql.get('variables'))

                request_data = {
                    'name': item.get('name', 'Unnamed Request'),
                    'url': url,
                    'method': req.get('method', 'POST'),
                    'headers': headers,
                    'query': query,
                    'variables': variables,
                    'operation_name': RequestImporter._extract_operation_name(query),
                    'folder': folder_path
                }

                requests.append(request_data)

            # Recursively process items
            if 'item' in item:
                current_folder = folder_path
                if item.get('name'):
                    current_folder = f"{folder_path}/{item['name']}" if folder_path else item['name']

                for sub_item in item['item']:
                    extract_requests(sub_item, current_folder)

        if 'item' in collection:
            for item in collection['item']:
                extract_requests(item)

        return requests
    
    @staticmethod
    def from_postman_collection(file_path: str) -> List[Dict[str, Any]]:
        """
        Import requests from Postman Collection v2.1 format
        
        Args:
            file_path: Path to Postman collection JSON file
            
        Returns:
            List of request dictionaries with url, headers, body, etc.
        """
        with open(file_path, 'r', encoding='utf-8') as f:
            collection = json.load(f)
        return RequestImporter._extract_postman_requests(collection)

    @staticmethod
    def from_postman_collection_data(collection: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Import requests from an already-loaded Postman collection."""
        if not isinstance(collection, dict):
            raise ValueError("Postman collection data must be a dictionary")
        return RequestImporter._extract_postman_requests(collection)
    
    @staticmethod
    def from_curl_command(curl_cmd: str) -> Dict[str, Any]:
        """
        Parse a cURL command and extract request details
        
        Args:
            curl_cmd: cURL command string
            
        Returns:
            Request dictionary with url, headers, body, etc.
        """
        tokens = shlex.split(curl_cmd)
        if tokens and tokens[0] == 'curl':
            tokens = tokens[1:]

        method = None
        url = ""
        headers = {}
        body_text = ""

        i = 0
        while i < len(tokens):
            token = tokens[i]

            if token in ('-X', '--request') and i + 1 < len(tokens):
                method = tokens[i + 1].upper()
                i += 2
                continue

            if token in ('-H', '--header') and i + 1 < len(tokens):
                header = tokens[i + 1]
                if ':' in header:
                    key, value = header.split(':', 1)
                    headers[key.strip()] = value.strip()
                i += 2
                continue

            if token in ('-d', '--data', '--data-raw', '--data-binary', '--data-ascii') and i + 1 < len(tokens):
                if body_text:
                    body_text += tokens[i + 1]
                else:
                    body_text = tokens[i + 1]
                i += 2
                continue

            if token == '--get':
                method = 'GET'
                i += 1
                continue

            if token.startswith('http://') or token.startswith('https://'):
                url = token
                i += 1
                continue

            i += 1

        query = None
        variables = None

        if body_text:
            try:
                body_json = json.loads(body_text)
                query = body_json.get('query', '')
                variables = RequestImporter._normalize_variables(body_json.get('variables'))
            except Exception:
                query = body_text

        if not method:
            method = 'POST' if body_text else 'GET'

        return {
            'name': 'Imported from cURL',
            'url': url,
            'method': method,
            'headers': headers,
            'query': query,
            'variables': variables,
            'operation_name': RequestImporter._extract_operation_name(query)
        }
    
    @staticmethod
    def from_json(file_path: str) -> Dict[str, Any]:
        """
        Import request from JSON file
        
        Expected format:
        {
            "url": "https://api.example.com/graphql",
            "headers": {"Authorization": "Bearer token"},
            "query": "mutation { ... }",
            "variables": {...},
            "operation_name": "MyMutation"
        }
        
        Args:
            file_path: Path to JSON file
            
        Returns:
            Request dictionary
        """
        with open(file_path, 'r', encoding='utf-8') as f:
            data = json.load(f)

        if not isinstance(data, dict):
            raise ValueError("JSON request file must contain an object at the root")
        
        return {
            'name': data.get('name', 'Imported from JSON'),
            'url': data.get('url', ''),
            'method': data.get('method', 'POST'),
            'headers': data.get('headers', {}),
            'query': data.get('query', ''),
            'variables': RequestImporter._normalize_variables(data.get('variables')),
            'operation_name': data.get('operation_name')
        }
    
    @staticmethod
    def from_yaml(file_path: str) -> Dict[str, Any]:
        """
        Import request from YAML file
        
        Expected format:
        url: https://api.example.com/graphql
        headers:
          Authorization: Bearer token
        query: |
          mutation { ... }
        variables:
          key: value
        
        Args:
            file_path: Path to YAML file
            
        Returns:
            Request dictionary
        """
        with open(file_path, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f)

        if not isinstance(data, dict):
            raise ValueError("YAML request file must contain an object at the root")
        
        return {
            'name': data.get('name', 'Imported from YAML'),
            'url': data.get('url', ''),
            'method': data.get('method', 'POST'),
            'headers': data.get('headers', {}),
            'query': data.get('query', ''),
            'variables': RequestImporter._normalize_variables(data.get('variables')),
            'operation_name': data.get('operation_name')
        }
    
    @staticmethod
    def from_raw_http(raw_request: str) -> Dict[str, Any]:
        """
        Parse raw HTTP request string
        
        Args:
            raw_request: Raw HTTP request string
            
        Returns:
            Request dictionary
        """
        normalized = raw_request.replace('\r\n', '\n').replace('\r', '\n').strip()
        lines = normalized.split('\n')
        
        # Parse request line
        request_line = lines[0]
        method_match = re.match(r'(\w+)\s+([^\s]+)', request_line)
        method = method_match.group(1) if method_match else 'POST'
        path = method_match.group(2) if method_match else '/'
        
        # Extract Host header to build full URL
        host = None
        headers = {}
        body_start = 0
        
        for i, line in enumerate(lines[1:], 1):
            if not line.strip():
                body_start = i + 1
                break
            
            if ':' in line:
                key, value = line.split(':', 1)
                headers[key.strip()] = value.strip()
                if key.strip().lower() == 'host':
                    host = value.strip()
        
        # Build URL
        protocol = 'https' if '443' in str(host) else 'http'
        if host and ':' in host:
            host, port = host.split(':')
            if port not in ['80', '443']:
                url = f"{protocol}://{host}:{port}{path}"
            else:
                url = f"{protocol}://{host}{path}"
        elif host:
            url = f"{protocol}://{host}{path}"
        else:
            url = path
        
        # Extract body
        body_text = '\n'.join(lines[body_start:]) if body_start < len(lines) else ""
        
        query = None
        variables = None
        
        if body_text:
            try:
                body_json = json.loads(body_text)
                query = body_json.get('query', '')
                variables = RequestImporter._normalize_variables(body_json.get('variables'))
            except Exception:
                query = body_text
        
        return {
            'name': 'Imported from raw HTTP',
            'url': url,
            'method': method,
            'headers': headers,
            'query': query,
            'variables': variables,
            'operation_name': RequestImporter._extract_operation_name(query)
        }
    
    @staticmethod
    def auto_detect_and_import(file_path: str) -> Any:
        """
        Auto-detect file format and import
        
        Args:
            file_path: Path to file
            
        Returns:
            Imported request(s) - list for Postman collections, dict for others
        """
        path = Path(file_path)
        suffix = path.suffix.lower()
        
        if suffix == '.json':
            # Try Postman collection first
            try:
                with open(file_path, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    if 'info' in data and 'schema' in data.get('info', {}):
                        # Looks like Postman collection
                        return RequestImporter.from_postman_collection(file_path)
            except:
                pass
            
            # Fall back to simple JSON
            return RequestImporter.from_json(file_path)
        
        elif suffix in ['.yaml', '.yml']:
            return RequestImporter.from_yaml(file_path)
        elif suffix in ['.txt', '.http', '.req', '.curl']:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read().strip()
            if content.startswith('curl '):
                return RequestImporter.from_curl_command(content)
            if re.match(r'^[A-Z]+\s+\S+\s+HTTP/\d\.\d$', content.splitlines()[0]):
                return RequestImporter.from_raw_http(content)
            raise ValueError(f"Unsupported text request format in file: {file_path}")
        
        else:
            with open(file_path, 'r', encoding='utf-8') as f:
                content = f.read().strip()
            if content.startswith('curl '):
                return RequestImporter.from_curl_command(content)
            if content and re.match(r'^[A-Z]+\s+\S+\s+HTTP/\d\.\d$', content.splitlines()[0]):
                return RequestImporter.from_raw_http(content)
            raise ValueError(f"Unsupported file format: {suffix}")

File: burp-extension/src/main/java/graphqlhunter/importer/RequestImporter.java
package graphqlhunter.importer;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
import graphqlhunter.GraphQLHunterJson;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class RequestImporter
{
    private static final ObjectMapper YAML_MAPPER = new ObjectMapper(new YAMLFactory());
    private static final Pattern HTTP_REQUEST_LINE = Pattern.compile("^([A-Z]+)\\s+(\\S+)\\s+HTTP/\\d\\.\\d$");
    private static final Pattern OPERATION_NAME_PATTERN = Pattern.compile("(?:query|mutation|subscription)\\s+(\\w+)");

    private RequestImporter()
    {
    }

    public static ImportedRequest fromCurlCommand(String curlCommand)
    {
        List<String> tokens = shellSplit(curlCommand);
        if (!tokens.isEmpty() && "curl".equals(tokens.getFirst()))
        {
            tokens = tokens.subList(1, tokens.size());
        }

        ImportedRequest request = new ImportedRequest();
        String bodyText = "";

        for (int index = 0; index < tokens.size(); index++)
        {
            String token = tokens.get(index);
            if (("-X".equals(token) || "--request".equals(token)) && index + 1 < tokens.size())
            {
                request.method = tokens.get(++index).toUpperCase(Locale.ROOT);
                continue;
            }
            if (("-H".equals(token) || "--header".equals(token)) && index + 1 < tokens.size())
            {
                String header = tokens.get(++index);
                int separator = header.indexOf(':');
                if (separator > 0)
                {
                    request.headers.put(header.substring(0, separator).trim(), header.substring(separator + 1).trim());
                }
                continue;
            }
            if (token.startsWith("http://") || token.startsWith("https://"))
            {
                request.url = token;
                continue;
            }
            if ((token.startsWith("-d") || token.startsWith("--data")) && index + 1 < tokens.size())
            {
                bodyText += tokens.get(++index);
            }
        }

        request.name = "Imported from cURL";
        if (request.method == null || request.method.isBlank())
        {
            request.method = bodyText.isBlank() ? "GET" : "POST";
        }
        applyGraphQlBody(request, bodyText);
        return request;
    }

    public static ImportedRequest fromRawHttp(String rawRequest)
    {
        String normalized = rawRequest.replace("\r\n", "\n").replace('\r', '\n').trim();
        String[] lines = normalized.split("\n");
        if (lines.length == 0)
        {
            throw new IllegalArgumentException("Raw HTTP request is empty");
        }

        Matcher matcher = HTTP_REQUEST_LINE.matcher(lines[0]);
        ImportedRequest request = new ImportedRequest();
        request.name = "Imported from raw HTTP";
        boolean matched = matcher.find();
        request.method = matched ? matcher.group(1) : "POST";
        String path = matched ? matcher.group(2) : "/";

        int bodyStart = lines.length;
        String host = "";
        for (int index = 1; index < lines.length; index++)
        {
            String line = lines[index];
            if (line.isBlank())
            {
                bodyStart = index + 1;
                break;
            }
            int separator = line.indexOf(':');
            if (separator > 0)
            {
                String key = line.substring(0, separator).trim();
                String value = line.substring(separator + 1).trim();
                request.headers.put(key, value);
                if ("host".equalsIgnoreCase(key))
                {
                    host = value;
                }
            }
        }

        request.url = buildUrl(host, path);
        String body = bodyStart >= lines.length ? "" : String.join("\n", java.util.Arrays.copyOfRange(lines, bodyStart, lines.length));
        applyGraphQlBody(request, body);
        return request;
    }

    public static ImportedRequest fromJsonContent(String json)
    {
        try
        {
            JsonNode node = GraphQLHunterJson.mapper().readTree(json);
            ImportedRequest request = new ImportedRequest();
            request.name = text(node, "name", "Imported from JSON");
            request.url = text(node, "url", "");
            request.method = text(node, "method", "POST");
            request.query = text(node, "query", "");
            request.operationName = text(node, "operation_name", text(node, "operationName", extractOperationName(request.query)));
            request.variables = node.has("variables") && !node.get("variables").isNull()
                ? GraphQLHunterJson.mapper().convertValue(node.get("variables"), Object.class)
                : new LinkedHashMap<String, Object>();
            if (node.has("headers") && node.get("headers").isObject())
            {
                node.get("headers").fields().forEachRemaining(entry -> request.headers.put(entry.getKey(), entry.getValue().asText("")));
            }
            return request;
        }
        catch (Exception exception)
        {
            throw new IllegalArgumentException("Unable to parse JSON request content", exception);
        }
    }

    public static ImportedRequest fromYamlContent(String yaml)
    {
        try
        {
            Object parsed = YAML_MAPPER.readValue(yaml, Object.class);
            return fromJsonContent(GraphQLHunterJson.write(parsed));
        }
        catch (Exception exception)
        {
            throw new IllegalArgumentException("Unable to parse YAML request content", exception);
        }
    }

    public static List<ImportedRequest> fromPostmanCollectionContent(String json)
    {
        try
        {
            JsonNode root = GraphQLHunterJson.mapper().readTree(json);
            List<ImportedRequest> requests = new ArrayList<>();
            JsonNode items = root.path("item");
            if (items.isArray())
            {
                for (JsonNode item : items)
                {
                    extractPostmanRequests(item, "", requests);
                }
            }
            return requests;
        }
        catch (Exception exception)
        {
            throw new IllegalArgumentException("Unable to parse Postman collection", exception);
        }
    }

    public static List<ImportedRequest> autoDetect(String fileName, String content)
    {
        String lowerName = fileName == null ? "" : fileName.toLowerCase(Locale.ROOT);
        String trimmed = content == null ? "" : content.trim();
        if (trimmed.startsWith("{") || trimmed.startsWith("["))
        {
            try
            {
                JsonNode node = GraphQLHunterJson.mapper().readTree(content);
                if ((node.has("info") && node.path("info").has("schema")) || node.has("item"))
                {
                    return fromPostmanCollectionContent(content);
                }
                return List.of(fromJsonContent(content));
            }
            catch (Exception ignored)
            {
            }
        }
        if (looksLikeYamlRequest(content))
        {
            try
            {
                return List.of(fromYamlContent(content));
            }
            catch (Exception ignored)
            {
            }
        }
        if (lowerName.endsWith(".json"))
        {
            try
            {
                JsonNode node = GraphQLHunterJson.mapper().readTree(content);
                if (node.has("info") && node.path("info").has("schema"))
                {
                    return fromPostmanCollectionContent(content);
                }
            }
            catch (Exception ignored)
            {
            }
            return List.of(fromJsonContent(content));
        }
        if (lowerName.endsWith(".yaml") || lowerName.endsWith(".yml"))
        {
            return List.of(fromYamlContent(content));
        }
        if (content.trim().startsWith("curl "))
        {
            return List.of(fromCurlCommand(content.trim()));
        }
        if (HTTP_REQUEST_LINE.matcher(content.split("\\R", 2)[0]).matches())
        {
            return List.of(fromRawHttp(content));
        }
        throw new IllegalArgumentException("Unsupported request format");
    }

    public static String extractOperationNameFromQuery(String query)
    {
        return extractOperationName(query);
    }

    private static void extractPostmanRequests(JsonNode item, String folderPath, List<ImportedRequest> requests)
    {
        if (item.has("request"))
        {
            JsonNode requestNode = item.get("request");
            ImportedRequest request = new ImportedRequest();
            request.name = text(item, "name", "Unnamed Request");
            request.folder = folderPath;
            request.method = text(requestNode, "method", "POST");
            request.url = buildPostmanUrl(requestNode.path("url"));
            JsonNode headers = requestNode.path("header");
            if (headers.isArray())
            {
                for (JsonNode header : headers)
                {
                    if (!header.path("disabled").asBoolean(false))
                    {
                        request.headers.put(text(header, "key", ""), text(header, "value", ""));
                    }
                }
            }

            JsonNode body = requestNode.path("body");
            if ("raw".equals(body.path("mode").asText("")))
            {
                applyGraphQlBody(request, body.path("raw").asText(""));
            }
            else if ("graphql".equals(body.path("mode").asText("")))
            {
                request.query = body.path("graphql").path("query").asText("");
                if (body.path("graphql").has("variables"))
                {
                    request.variables = normalizeVariables(body.path("graphql").get("variables"));
                }
                request.operationName = extractOperationName(request.query);
            }

            requests.add(request);
        }

        JsonNode nestedItems = item.path("item");
        if (nestedItems.isArray())
        {
            String nextFolder = folderPath;
            if (item.has("name"))
            {
                nextFolder = folderPath.isBlank() ? item.get("name").asText("") : folderPath + "/" + item.get("name").asText("");
            }
            for (JsonNode nested : nestedItems)
            {
                extractPostmanRequests(nested, nextFolder, requests);
            }
        }
    }

    private static void applyGraphQlBody(ImportedRequest request, String body)
    {
        if (body == null || body.isBlank())
        {
            return;
        }
        try
        {
            JsonNode bodyJson = GraphQLHunterJson.mapper().readTree(body);
            if (bodyJson.isObject())
            {
                request.query = bodyJson.path("query").asText(body);
                if (bodyJson.has("variables"))
                {
                    request.variables = normalizeVariables(bodyJson.get("variables"));
                }
                request.operationName = bodyJson.has("operationName")
                    ? bodyJson.get("operationName").asText("")
                    : extractOperationName(request.query);
                return;
            }
        }
        catch (Exception ignored)
        {
        }
        request.query = body;
        request.operationName = extractOperationName(body);
    }

    private static Object normalizeVariables(JsonNode variables)
    {
        if (variables == null || variables.isNull())
        {
            return new LinkedHashMap<String, Object>();
        }
        if (variables.isTextual())
        {
            String raw = variables.asText("");
            if (raw.isBlank())
            {
                return new LinkedHashMap<String, Object>();
            }
            try
            {
                return GraphQLHunterJson.mapper().readValue(raw, Object.class);
            }
            catch (Exception exception)
            {
                return raw;
            }
        }
        return GraphQLHunterJson.mapper().convertValue(variables, Object.class);
    }

    private static String extractOperationName(String query)
    {
        if (query == null || query.isBlank())
        {
            return "";
        }
        Matcher matcher = OPERATION_NAME_PATTERN.matcher(query);
        return matcher.find() ? matcher.group(1) : "";
    }

    private static boolean looksLikeYamlRequest(String content)
    {
        String lowered = content == null ? "" : content.toLowerCase(Locale.ROOT);
        return lowered.contains("query:")
            || lowered.contains("url:")
            || lowered.contains("method:")
            || lowered.contains("variables:")
            || lowered.contains("headers:");
    }

    private static String text(JsonNode node, String field, String defaultValue)
    {
        return node.has(field) ? node.get(field).asText(defaultValue) : defaultValue;
    }

    private static String buildPostmanUrl(JsonNode urlNode)
    {
        if (urlNode.isTextual())
        {
            return urlNode.asText("");
        }
        String protocol = text(urlNode, "protocol", "https");
        String host = joinArray(urlNode.path("host"), ".");
        String path = joinArray(urlNode.path("path"), "/");
        return (protocol + "://" + host + "/" + path).replace("//", "/").replace(":/", "://");
    }

    private static String joinArray(JsonNode node, String separator)
    {
        if (!node.isArray())
        {
            return node.asText("");
        }
        List<String> parts = new ArrayList<>();
        node.forEach(value -> parts.add(value.asText("")));
        return String.join(separator, parts);
    }

    private static String buildUrl(String host, String path)
    {
        if (host == null || host.isBlank())
        {
            return path;
        }
        String protocol = host.contains(":443") ? "https" : "http";
        if (!host.contains(":"))
        {
            return protocol + "://" + host + path;
        }
        try
        {
            URI uri = new URI(protocol + "://" + host);
            return uri.toString() + path;
        }
        catch (URISyntaxException exception)
        {
            return protocol + "://" + host + path;
        }
    }

    private static List<String> shellSplit(String command)
    {
        ArrayList<String> tokens = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean single = false;
        boolean dbl = false;
        for (int index = 0; index < command.length(); index++)
        {
            char character = command.charAt(index);
            if (character == '\'' && !dbl)
            {
                single = !single;
                continue;
            }
            if (character == '"' && !single)
            {
                dbl = !dbl;
                continue;
            }
            if (Character.isWhitespace(character) && !single && !dbl)
            {
                if (!current.isEmpty())
                {
                    tokens.add(current.toString());
                    current.setLength(0);
                }
                continue;
            }
            current.append(character);
        }
        if (!current.isEmpty())
        {
            tokens.add(current.toString());
        }
        return tokens;
    }
}
